src/ucp/tag/tag_match.inl

```diff
@@ -212,29 +213,96 @@ ucp_tag_unexp_search(ucp_tag_match_t *tm, ucp_tag_t tag, uint64_t tag_mask,
     return NULL;
 }
 
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_request_recv_offload_data(ucp_request_t *req, const void *data,
+                              size_t length, size_t offset, unsigned recv_flags,
+                              int *last)
+{
+    ucs_status_t status = UCS_OK;
+    ucp_offload_last_ssend_hdr_t *s_hdr;
+
+    /* Should be used in multi-fragmented flow only */
+    ucs_assert(!(recv_flags & UCP_RECV_DESC_FLAG_EAGER_ONLY));
+
+    if (recv_flags & UCP_RECV_DESC_FLAG_EAGER_LAST) {
+        *last = 1;
+        if (recv_flags & UCP_RECV_DESC_FLAG_EAGER_SYNC) {
+            s_hdr = (ucp_offload_last_ssend_hdr_t*)UCS_PTR_BYTE_OFFSET(data,
+                                                                       -sizeof(*s_hdr));
+            ucp_tag_offload_sync_send_ack(req->recv.worker,
+                                          s_hdr->ssend_ack.ep_ptr,
+                                          s_hdr->ssend_ack.sender_tag,
+                                          recv_flags);
+        }
+    } else {
+        *last = 0;
+    }
+
+    if (ucs_unlikely(req->status != UCS_OK)) {
+       return req->status;
+    }
+
+    if (ucs_unlikely(req->recv.length < length + offset)) {
+        return ucp_request_recv_msg_truncated(req, length, offset);
+    }
+
+    if (UCP_DT_IS_CONTIG(req->recv.datatype)) {
+        ucp_request_unpack_contig(req, req->recv.buffer + offset, data, length);
+    } else {
+        /* For non-contig data need to assemble the whole message
+         * before calling unpack. */
+        if (offset == 0) {
+            req->recv.tag.gen_buf = ucs_malloc(req->recv.length, "tag gen buffer");
```

Comment:
can we allocate from one of the mpools on rx side?
anyway the size is limited by RNDV_THREHS right?
or at least allocate from mpool for small enough messages